File: LAB3-IMU VectorNav-VN100/analysis/allan.m
% Load CSV data
data = readtable('/home/shrirag10/LAB3/src/imu_data.csv');  % Replace with your actual file path

% Extract accelerometer and gyroscope data
accel_x = data.accel_x;  % Adjust column names as per your CSV
accel_y = data.accel_y;  % Adjust column names as per your CSV
accel_z = data.accel_z;  % Adjust column names as per your CSV
gyro_x = data.gyro_x;    % Adjust column names as per your CSV
gyro_y = data.gyro_y;    % Adjust column names as per your CSV
gyro_z = data.gyro_z;    % Adjust column names as per your CSV

% Combine the accelerometer and gyroscope axes data
accel_combined = sqrt(accel_x.^2 + accel_y.^2 + accel_z.^2);
gyro_combined = sqrt(gyro_x.^2 + gyro_y.^2 + gyro_z.^2);

% Sampling time interval (assuming uniform sampling)
dt = 1 / 100;  % Adjust based on your actual sampling rate

% Function to compute Allan deviation
function [taus, allan_deviation] = compute_allan_deviation(data, dt)
    N = length(data);   % Number of data points
    max_tau = floor(N / 2);  % Maximum tau is half the data length
    taus = logspace(log10(1), log10(max_tau), 50);  % Tau values spaced logarithmically
    
    allan_deviation = zeros(size(taus));  % Initialize Allan deviation array
    
    for i = 1:length(taus)
        tau = round(taus(i));  % Ensure tau is an integer
        num_averages = floor(N / tau);  % Number of averages for this tau
        
        if num_averages < 2
            allan_deviation(i) = NaN;  % Skip computation for invalid values
            continue;
        end

        % Compute the averages of the data over non-overlapping windows of size tau
        avg_values = arrayfun(@(j) mean(data((j-1)*tau + 1 : j*tau)), 1:num_averages);
        
        % Compute the Allan variance
        allan_var = 0.5 * mean(diff(avg_values).^2);
        
        % Store Allan deviation (sqrt of variance)
        allan_deviation(i) = sqrt(allan_var);
    end
end

% Calculate Allan deviation for combined accelerometer and gyroscope data
[tau_accel, allan_dev_accel] = compute_allan_deviation(accel_combined, dt);
[tau_gyro, allan_dev_gyro] = compute_allan_deviation(gyro_combined, dt);

% Filter valid tau and Allan deviation values for accelerometer
valid_idx_accel = ~isnan(allan_dev_accel) & allan_dev_accel > 0;

% Plot Allan deviation for Accelerometer
figure;
loglog(tau_accel(valid_idx_accel), allan_dev_accel(valid_idx_accel), 'b', 'LineWidth', 2);
xlabel('\tau (s)');
ylabel('\sigma(\tau) (m/s^2)');
title('Allan Deviation for Accelerometer');
grid on;

% Filter valid tau and Allan deviation values for gyroscope
valid_idx_gyro = ~isnan(allan_dev_gyro) & allan_dev_gyro > 0;

% Plot Allan deviation for Gyroscope
figure;
loglog(tau_gyro(valid_idx_gyro), allan_dev_gyro(valid_idx_gyro), 'g', 'LineWidth', 2);
xlabel('\tau (s)');
ylabel('\sigma(\tau) (rad/s)');
title('Allan Deviation for Gyroscope');
grid on;


%Debbugging
disp('First 5 rows of accelerometer combined data:');
disp(accel_combined(1:5));

disp('First 5 rows of gyroscope combined data:');
disp(gyro_combined(1:5));

disp('Tau values for accelerometer:');
disp(tau_accel(1:5));

disp('Allan deviation for accelerometer:');
disp(allan_dev_accel(1:5));

disp('Tau values for gyroscope:');
disp(tau_gyro(1:5));

disp('Allan deviation for gyroscope:');
disp(allan_dev_gyro(1:5));
% Plot raw combined accelerometer data
figure;
plot(accel_combined);
title('Raw Accelerometer Data');
xlabel('Sample Index');
ylabel('Acceleration (m/s^2)');
grid on;

% Plot raw combined gyroscope data
figure;
plot(gyro_combined);
title('Raw Gyroscope Data');
xlabel('Sample Index');
ylabel('Angular Velocity (rad/s)');
grid on;

File: LAB3-IMU VectorNav-VN100/analysis/allan.py
import matplotlib.pyplot as plt
import numpy as np
from pathlib import Path
from rosbags.highlevel import AnyReader
from rosbags.typesys import Stores, get_typestore, get_types_from_msg
from tf_transformations import euler_from_quaternion

# Load ROS2 bag data for IMU sensor readings
def load_ros2_bag_data(bagfile):
    imu_data = {'time': [], 'accel': [], 'gyro': [], 'orientation': []}
    
    bagpath = Path(bagfile)
    msg_text = Path('imu_msgs/msg/IMUmsg.msg').read_text()

    # Create a type store
    typestore = get_typestore(Stores.ROS2_HUMBLE)
    add_types = {}
    add_types.update(get_types_from_msg(msg_text, 'imu_msgs/msg/IMUmsg'))

    typestore.register(add_types)
    
    # Reading the data from the bagfile
    with AnyReader([bagpath], default_typestore=typestore) as reader:
        connections = [x for x in reader.connections if x.topic == '/imu']
        for connection, timestamp, rawdata in reader.messages(connections=connections):
            msg = reader.deserialize(rawdata, connection.msgtype)
            
            # Extract accelerometer and gyroscope data
            accel_x = msg.imu.linear_acceleration.x
            accel_y = msg.imu.linear_acceleration.y
            accel_z = msg.imu.linear_acceleration.z

            gyro_x = msg.imu.angular_velocity.x
            gyro_y = msg.imu.angular_velocity.y
            gyro_z = msg.imu.angular_velocity.z

            # Append extracted data to the dictionary
            imu_data['time'].append(timestamp / 1e9)  # Convert nanoseconds to seconds
            imu_data['accel'].append([accel_x, accel_y, accel_z])
            imu_data['gyro'].append([gyro_x, gyro_y, gyro_z])
            imu_data['orientation'].append([msg.imu.orientation.x, msg.imu.orientation.y, msg.imu.orientation.z, msg.imu.orientation.w])
    
    return imu_data

# Convert quaternions to Euler angles
def convert_to_euler(orientations):
    euler_angles = []
    for quat in orientations:
        roll, pitch, yaw = euler_from_quaternion(quat)
        euler_angles.append([roll, pitch, yaw])
    return np.array(euler_angles)

# Compute Allan deviation
def compute_allan_deviation(data, dt):
    N = len(data)
    max_tau = N // 2
    tau_values = np.logspace(np.log10(1), np.log10(max_tau), num=50).astype(int)

    allan_dev = []
    for tau in tau_values:
        avgs = [np.mean(data[i:i + tau]) for i in range(0, N - tau, tau)]
        adev = np.sqrt(0.5 * np.mean(np.diff(avgs) ** 2))
        allan_dev.append(adev)
    
    return tau_values * dt, np.array(allan_dev)  # Convert to NumPy array

# Function to plot Allan deviation for accelerometer and gyroscope separately
def plot_separate_allan_variance(tau, allan_dev_accel, allan_dev_gyro):
    # Plotting for Accelerometer
    plt.figure(figsize=(10, 8))  # Adjust figure size for better visibility
    plt.loglog(tau, allan_dev_accel, label='Combined Accelerometer Axes', color='b')

    # Set axis limits for accelerometer
    plt.xlim(1e-2, 1e6)
    plt.ylim(1e-4, 1e4)

    plt.title("Allan Deviation for Accelerometer", fontsize=14)
    plt.xlabel(r'$\tau$ (s)', fontsize=12)
    plt.ylabel(r'$\sigma(\tau)$', fontsize=12)

    plt.grid(True, which="both", ls="--", lw=0.5)  # Dotted lines for grid
    plt.minorticks_on()
    plt.legend(loc='best')
    plt.show()

    # Plotting for Gyroscope
    plt.figure(figsize=(10, 8))  # Adjust figure size for better visibility
    plt.loglog(tau, allan_dev_gyro, label='Combined Gyroscope Axes', color='g')

    # Set axis limits for gyroscope
    plt.xlim(1e-2, 1e6)

    # Adjust the y-axis limits based on the observed data range for gyroscope
    min_y_gyro = np.min(allan_dev_gyro[allan_dev_gyro > 0]) * 0.1  # Set a minimum limit based on data
    max_y_gyro = np.max(allan_dev_gyro) * 10  # Set a maximum limit based on data
    plt.ylim(min_y_gyro, max_y_gyro)

    plt.title("Allan Deviation for Gyroscope", fontsize=14)
    plt.xlabel(r'$\tau$ (s)', fontsize=12)
    plt.ylabel(r'$\sigma(\tau)$', fontsize=12)

    plt.grid(True, which="both", ls="--", lw=0.5)
    plt.minorticks_on()
    plt.legend(loc='best')
    plt.show()

# Combine accelerometer and gyroscope axes data
def combine_axes_data(imu_data, sensor_type):
    combined_data = np.linalg.norm(np.array(imu_data[sensor_type]), axis=1)
    return combined_data

# Main function to analyze ros2 bag data
def analyze_imu_data(db3_file):
    imu_data = load_ros2_bag_data(db3_file)
    
    # Sampling time interval (assuming uniform sampling)
    dt = np.mean(np.diff(imu_data['time']))
    
    # Combine X, Y, Z axes for accelerometer and gyroscope
    combined_accel_data = combine_axes_data(imu_data, 'accel')
    combined_gyro_data = combine_axes_data(imu_data, 'gyro')

    # Calculate Allan deviation for combined accelerometer and gyroscope data
    tau, allan_dev_accel = compute_allan_deviation(combined_accel_data, dt)
    tau, allan_dev_gyro = compute_allan_deviation(combined_gyro_data, dt)

    # Plot separate Allan variance for accelerometer and gyroscope
    plot_separate_allan_variance(tau, allan_dev_accel, allan_dev_gyro)

# Example usage
if __name__ == "__main__":
    bag_file = 'imu_driver/fivehour1data.bag'  # Replace with your actual file path
    analyze_imu_data(bag_file)

